### PR TITLE
Add jupyter notebook sandbox example

### PR DESCRIPTION
**Pull request by pawalt**
This PR adds a new sandbox example that shows how to spin up a Jupyter notebook in a sandbox.

### Type of Change

- [x] New example
- [ ] Example updates (Bug fixes, new features, etc.)
- [ ] Other (changes to the codebase, but not to examples)

## Checklist

- [x] Example is testable in synthetic monitoring system, or `lambda-test: false` is added to example frontmatter (`---`)
  - [x] Example is tested by executing with `modal run` or an alternative `cmd` is provided in the example frontmatter (e.g. `cmd: ["modal", "deploy"]`)
  - [x] Example is tested by running with no arguments or the `args` are provided in the example frontmatter (e.g. `args: ["--prompt", "Formula for room temperature superconductor:"]`
- [x] Example is documented with comments throughout, in a [_Literate Programming_](https://en.wikipedia.org/wiki/Literate_programming) style.
- [x] Example does _not_ require third-party dependencies to be installed locally
- [x] Example pins its dependencies
  - [x] Example pins container images to a stable tag, not a dynamic tag like `latest`
  - [x] Example specifies a `python_version` for the base image, if it is used
  - [x] Example pins all dependencies to at least minor version, `~=x.y.z` or `==x.y`
  - [x] Example dependencies with `version < 1` are pinned to patch version, `==0.y.z`

### CODE REVIEW

**Comment by charlesfrye** [comment]
🚀 The docs preview is ready! Check it out here:     https://modal-labs-examples--frontend-preview-f07de40.modal.run

**Comment by charlesfrye** [comment]
🚀 The docs preview is ready! Check it out here:     https://modal-labs-examples--frontend-preview-39d6656.modal.run

**Comment by charlesfrye** [comment]
🚀 The docs preview is ready! Check it out here:     https://modal-labs-examples--frontend-preview-29f83e3.modal.run

**Comment by charlesfrye** [comment]
🚀 The docs preview is ready! Check it out here:     https://modal-labs-examples--frontend-preview-edcb5e7.modal.run

File: 13_sandboxes/jupyter_sandbox.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+# Finally, we can print out a URL that we can use to connect to our Jupyter server.
+# Note that we have to call [`Sandbox.tunnels`](/docs/reference/modal.Sandbox#tunnels) to get the
+# URL because the Sandbox is not publicly accessible until we do so.
+
+tunnel = sandbox.tunnels()[JUPYTER_PORT]
+url = f"{tunnel.url}/?token={token}"
+print(f"Jupyter notebook is running at: {url}")
+
+# You can now open this URL in your browser to access the Jupyter notebook! When you're done,
```
Comment (charlesfrye) [inline 13_sandboxes/jupyter_sandbox.py:66]: Would love a line of code or two to test this -- a quick and cutty manual urllib request to a Jupyter API endpoint that demonstrates the server is live.

More for testing than for users.

File: 13_sandboxes/jupyter_sandbox.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+# terminate the sandbox using the UI or `Sandbox.from_id(sandbox.object_id).terminate()`.
+
+sandbox.terminate()
```
Comment (charlesfrye) [inline 13_sandboxes/jupyter_sandbox.py:69]: Doesn't this close the Jupyter server basically immediately? Would prefer a way for users to check on the server themselves real quick. A timeout maybe?

File: 13_sandboxes/jupyter_sandbox.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
+# ---
+# cmd: ["python", "13_sandboxes/jupyter_sandbox.py"]
+# tags: ["use-case-sandboxed-code-execution"]
+# ---
+
+# # Run a Jupyter notebook in a Modal Sandbox
+#
+# This example demonstrates how to run a Jupyter notebook in a Modal [Sandbox](/docs/guide/sandbox).
+
+import secrets
```
Comment (charlesfrye) [inline 13_sandboxes/jupyter_sandbox.py:10]: cool lib, TIL

File: 13_sandboxes/jupyter_sandbox.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+# All Sandboxes are associated with an App. We look up our app by name, creating it if it doesn't exist.
+
+app = modal.App.lookup("example-jupyter", create_if_missing=True)
+
+# We define a custom Docker image that has Jupyter and some other dependencies installed.
+# Using a custom image allows us to avoid re-installing these deps on every Sandbox startup.
+
+image = (
+    modal.Image.debian_slim()
```
Comment (charlesfrye) [inline 13_sandboxes/jupyter_sandbox.py:22]: could you set the Python version here? 3.12 probably

File: 13_sandboxes/jupyter_sandbox.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    .pip_install("jupyter~=1.1.0", "bing-image-downloader~=1.1.2")
```
Comment (charlesfrye) [inline 13_sandboxes/jupyter_sandbox.py:23]: Do we need the bing image downloader?

File: 13_sandboxes/jupyter_sandbox.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    .pip_install("pandas~=2.2.3")  # Any other deps
```
Comment (charlesfrye) [inline 13_sandboxes/jupyter_sandbox.py:24]: I like the callout to "other deps", but I don't think we need a separate pip install.

Maybe a commented-out `.pip_install(pandas, numpy, seaborn)`?

**Review by charlesfrye — CHANGES_REQUESTED** [review]

**Comment by charlesfrye** [comment]
🚀 The docs preview is ready! Check it out here:     https://modal-labs-examples--frontend-preview-0dd309e.modal.run

File: 13_sandboxes/jupyter_sandbox.py (hunk 2 above)
Comment (pawalt) [reply 13_sandboxes/jupyter_sandbox.py:69]: sure yeah i can do a timeout. i just did this for explicit resource cleanup

File: 13_sandboxes/jupyter_sandbox.py (hunk 5 above)
Comment (pawalt) [reply 13_sandboxes/jupyter_sandbox.py:23]: no

File: 13_sandboxes/jupyter_sandbox.py (hunk 6 above)
Comment (pawalt) [reply 13_sandboxes/jupyter_sandbox.py:24]: ya good call

File: 13_sandboxes/jupyter_sandbox.py (hunk 1 above)
Comment (pawalt) [reply 13_sandboxes/jupyter_sandbox.py:93]: done

**Comment by charlesfrye** [comment]
🚀 The docs preview is ready! Check it out here:     https://modal-labs-examples--frontend-preview-0a28318.modal.run

**Comment by charlesfrye** [comment]
🚀 The docs preview is ready! Check it out here:     https://modal-labs-examples--frontend-preview-8d91286.modal.run

**Comment by charlesfrye** [comment]
🚀 The docs preview is ready! Check it out here:     https://modal-labs-examples--frontend-preview-911e5ac.modal.run

**Comment by charlesfrye** [comment]
🚀 The docs preview is ready! Check it out here:     https://modal-labs-examples--frontend-preview-86a9b4a.modal.run

**Comment by charlesfrye** [comment]
🚀 The docs preview is ready! Check it out here:     https://modal-labs-examples--frontend-preview-6cceb08.modal.run

**Comment by charlesfrye** [comment]
🚀 The docs preview is ready! Check it out here:     https://modal-labs-examples--frontend-preview-dfce411.modal.run

**Comment by charlesfrye** [comment]
🚀 The docs preview is ready! Check it out here:     https://modal-labs-examples--frontend-preview-23cdc5b.modal.run

**Comment by charlesfrye** [comment]
🚀 The docs preview is ready! Check it out here:     https://modal-labs-examples--frontend-preview-15be389.modal.run

**Comment by charlesfrye** [comment]
🚀 The docs preview is ready! Check it out here:     https://modal-labs-examples--frontend-preview-79889fd.modal.run

**Comment by charlesfrye** [comment]
🚀 The docs preview is ready! Check it out here:     https://modal-labs-examples--frontend-preview-14b64ad.modal.run

**Comment by charlesfrye** [comment]
🚀 The docs preview is ready! Check it out here:     https://modal-labs-examples--frontend-preview-e83b2f7.modal.run

**Comment by charlesfrye** [comment]
🚀 The docs preview is ready! Check it out here:     https://modal-labs-examples--frontend-preview-21973cc.modal.run

**Review by charlesfrye — APPROVED** [review]

**Pull request merged**